sdk/core/azure-core-amqp/src/impl/rust_amqp/amqp/message_receiver.cpp

```diff
@@ -22,39 +28,85 @@
+      RustAmqpMessageReceiver* value)
   {
-    messagereceiver_destroy(value);
+    amqpmessagereceiver_destroy(value);
   }
-#endif
+
+  template <> struct UniqueHandleHelper<RustInterop::_detail::RustAmqpMessageReceiverOptions>
+  {
+    static void FreeMessageReceiverOptions(RustAmqpMessageReceiverOptions* obj);
+
+    using type = Core::_internal::
+        BasicUniqueHandle<RustAmqpMessageReceiverOptions, FreeMessageReceiverOptions>;
+  };
+
+  void UniqueHandleHelper<RustAmqpMessageReceiverOptions>::FreeMessageReceiverOptions(
+      RustInterop::_detail::RustAmqpMessageReceiverOptions* value)
+  {
+    amqpmessagereceiveroptions_destroy(value);
+  }
+
 }}}} // namespace Azure::Core::Amqp::_detail
 
 namespace Azure { namespace Core { namespace Amqp { namespace _detail {
+  using UniqueMessageReceiverOptions
+      = UniqueHandle<RustInterop::_detail::RustAmqpMessageReceiverOptions>;
 
   /** Configure the MessageReceiver for receiving messages from a service instance.
    */
   MessageReceiverImpl::MessageReceiverImpl(
       std::shared_ptr<_detail::SessionImpl> session,
       Models::_internal::MessageSource const& source,
       MessageReceiverOptions const& options)
-      : m_options{options}, m_source{source}, m_session{session}
+      : m_receiver{amqpmessagereceiver_create()}, m_options{options}, m_source{source}, m_session{
+                                                                                            session}
   {
   }
 
   std::pair<std::shared_ptr<Models::AmqpMessage>, Models::_internal::AmqpError>
   MessageReceiverImpl::WaitForIncomingMessage(Context const& context)
   {
-    throw std::runtime_error("Not implemented");
-    (void)context;
+    Common::_detail::CallContext callContext(
+        Common::_detail::GlobalStateHolder::GlobalStateInstance()->GetRuntimeContext(), context);
+
+    {
+
+      auto message = Models::_detail::AmqpMessageFactory::FromImplementation(
+          amqpmessagereceiver_receive_message_wait(callContext.GetCallContext(), m_receiver.get()));
+
+      return std::make_pair(message, Models::_internal::AmqpError{});
+    }
   }
+
   std::pair<std::shared_ptr<Models::AmqpMessage>, Models::_internal::AmqpError>
   MessageReceiverImpl::TryWaitForIncomingMessage()
   {
-    throw std::runtime_error("Not implemented");
+    Common::_detail::CallContext callContext(
+        Common::_detail::GlobalStateHolder::GlobalStateInstance()->GetRuntimeContext(), {});
+
+    auto message = Models::_detail::AmqpMessageFactory::FromImplementation(
+        amqpmessagereceiver_receive_message_async_poll(
+            callContext.GetCallContext(), m_receiver.get()));
+
+    if (message)
+    {
+      return std::make_pair(message, Models::_internal::AmqpError{});
+    }
+    else if (callContext.GetError().empty())
+    {
+      return std::make_pair(nullptr, Models::_internal::AmqpError{});
+    }
+    else
+    {
+      return std::make_pair(nullptr, Models::_internal::AmqpError{});
```

Comment:
Shouldn't this return an non-generic error? It returns the same thing as line 104 above.

Reply:
I have an outstanding work item (literally the next one on my list after I fix the two-stage receive pattern) to rationalize the Rust AMQP error story.

Part of that fix will involve hooking up the Rust AMQP errors into something meaningful to C++, which currently only understands the AMQP protocol level errors and none of the networking related errors.

The semantics for AmqpError is that a default constructed AmqpError indicates "no error" - so this case is saying "I don't have a message AND there is no error" - in other words, the poll operation succeeded but there was no message to be received.